text/0008-weighted-uniform-random-tip-selection/0008-weighted-uniform-random-tip-selection.md

```diff
@@ -0,0 +1,190 @@
+which means that transactions of which their `OTRSI` is more than 15, are "below max depth".
+
+### OTRSI / YTRSI example
+Given the blue PoV transaction, the `OTRSI` of it is milestone 1 and `YTRSI` milestone 2. The orange
+and purple transactions are the `Confirmed Root Transactions`.
+![sdf](images/otrsi_ytrsi.PNG)
+
+### Milestone based tip scoring
+
+The milestone based scoring defines a tip's score by investigating the tip's relation to
+the cone it approves and previous issued  milestones.
+
+A tip can have one of 3 score states:
+* `0`: The tip is lazy and should not be selected.
+* `1`: The tip is somewhat lazy.
+* `2`: The tip is a non-lazy tip.
+
+Definitions:
+* `C1`: Max allowed delta value for the `YTRSI` of a given transaction in relation to the current `LSMI`.
+* `C2`: Max allowed delta value between `OTRSI` of the approvees of a given transaction in relation to the current `LSMI`. 
+* `M`: Max allowed delta value between `OTRSI` of the given transaction in relation to the current `LSMI`.
+`M` is the `below max depth (BMD)` parameter.
+
+Recommended defaults:
+* `C1` = 2 milestones
+* `C2` = 7 milestones
+* `M` = 15 milestones
+
+Scoring Algorithm (pseudo code):
+```
+
+enum Score (
+    LAZY = 0
+    SEMI_LAZY = 1
+    NON_LAZY = 2
+)
+
+const (
+    C1 = 2
+    C2 = 7
+    M = 15
+)
+
+func score(tip Tip) Score {
+    
+    // if the LSMI to YTRSI delta is over C1, then the tip is lazy
+    if (LSMI - YTRSI(tip) > C1) {
+        return Score.LAZY
+    }
+    
+    // if the OTRSI to LSMI delta is over M/below-max-depth, then the tip is lazy
+    if (LSMI - OTRSI(tip) > M) {
+        return Score.LAZY
+    }
+    
+    // the approvees (trunk and branch) are the tail transactions this tip approves
+    approvees := tip.Approvees()
+    approveesOTRSICheck := 2
+    for (i := 0; i < 2; i++) {
+        approvee := approvees[i]
+    
+        // direct approvee is already lazy, therefore so is this tip
+        if (approvee.Score == 0) {
+            return Score.LAZY
+        }
+        
+        // if the OTRSI to LSMI delta of the approvee is C2, we mark it as such
+        if (LSMI - OTRSI(approvee) > C2) {
+            approveesOTRSICheck--
+        }
+    }
+
+    // if both approvees' OTRSI violates the LSMI delta in relation to C2 the tip is lazy too
+    if (approveesOTRSICheck == 0) {
+        return Score.LAZY
+    }
+    
+    // if only one of the approvees violated the OTRSI to LMSI delta, the tip is considered semi-lazy
+    if (approveesOTRSICheck == 1) {
+        return Score.SEMI_LAZY
+    }
+
+    return Score.NON_LAZY
+}
+```
+
+### Weighted Random Tip-Selection
+
+Given the scoring, a node should keep a set of tips with their associated score.
+
+Tip-Selection (pseudo code):
+```
+
+var tips = Set(tips_and_score)
+
+func select() Tip {
+    // compute the sum of the score of all tips
+    scoreSum := tips.ScoreSum()
+    
+    // get a random number between 1 and the score sum
+    r := rand.Int(1, scoreSum)

```

Comment:
This will fail, if there are only lazy tips as then `scoreSum == 0`. What should we do in that particular case?

Although this is a edge-case, it could still happen when the network connections are lost and only the coordinator was able to issue milestones in the recent past.

Reply:
If there are only lazy tips, then `tips` is empty and we return null (I've added a corresponding code snippet to catch that condition.). If the node is not synchronized or has no network connection, it should not execute tip-selection.

Reply:
I agree. Since it has now made clear that `Confirmed Root Transactions` also contain milestones, we will always at least have the milestone as a non-lazy tip for a synched node